Add --fail-no-examples option: fail if no RSpec examples

xmik opened the pull request:
I added a new command line option: `--fail-no-examples`. It defaults to false, so default behavior is not changed. But, if this option is set and there are no RSpec examples to be run, exit status will be 1 (so far it was always 0).

### Motivation
We use RSpec rake task on CI. Having updated RSpec from 3.3.0 to 3.5.0, there was some change in that rake task code. Without changing anything, the rake task resulted in running 0 examples and exit status was 0. Jobs were passed on CI, but I'd prefer if they were not.

myronmarston commented:
> Having updated RSpec from 3.3.0 to 3.5.0, there was some change in that rake task code.

Was there a regression in our rake task?  If so, it'd be good for us to know about it so we can fix it.

Anyhow, if we're going to add this, I think it would be better to add it to `RSpec::Core::Configuration` (so it can be set via `RSpec.configure`) rather than adding it as a CLI option.  We intentionally limit what config options are exposed via the CLI.  If we exposed everything, `rspec --help` would be an overwhelming wall of text.  We just want options that are the main ones users commonly want to run differently in a one-off manner.  This sounds like an option that you'd want to permanently set for your CI environment, not something for a one-off run.  If we added it to `RSpec.configure`, you could set it in your CI environment like so:

``` ruby
# spec/spec_helper.rb
RSpec.configure do |c|
  if ENV['CI']
    c.fail_if_no_examples = true
  end
end
```

```
# .rspec
--require spec_helper
```

(The latter is necessary to ensure `spec_helper.rb` is always loaded. Simply requiring it from spec files won't help if no spec files are loaded!)

Thoughts?

xmik commented:
Yes, you are right, I want to set this option permanently and setting it by `RSpec::Core::Configuration` is enough. I'm going to rebase this PR so that there is no CLI option, only `RSpec::Core::Configuration` one.

As it comes for rake task potential regression, I opened an issue: https://github.com/rspec/rspec-core/issues/2303 .

JonRowe commented on lib/rspec/core/configuration.rb:
```diff
@@ -199,6 +199,10 @@ def only_failures_but_not_configured?
 
+      # @macro add_setting
+      # Whether or not to fail when there are no RSpec examples (default: false).
+      add_setting :fail_no_examples
```
Currently the build is failing due to a drop in test coverage, I believe this needs a test covering it to fix that :)

xmik commented:
I've just rebased this PR. The option is named: `fail_if_no_examples` (as @myronmarston wrote in usage example) and added better tests, but let me know if not enough. I tried to test exit status in unit tests, but it always resulted in ruby errors (my fault), so I left exit status tested only in integration tests (cucumber).

And thanks for that comment on `.rspec` file, I understood that later.

myronmarston commented on lib/rspec/core/runner.rb:
```diff
@@ -110,6 +110,9 @@ def setup(err, out)
       def run_specs(example_groups)
         @configuration.reporter.report(@world.example_count(example_groups)) do |reporter|
           @configuration.with_suite_hooks do
+            if @world.example_count(example_groups) == 0 && @configuration.fail_if_no_examples
```
`@world.example_count(example_groups)` is being called above on line 111 and then again here on line 113.  Instead of calling it twice, can you call it once, store the value in a local variable, and then use it in two places?

xmik replied:
Yes, of course! I was experimenting here and forgot.

myronmarston commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,106 @@
+     end
+     """
+   When I run `rspec spec/example_spec.rb`
+   Then the exit status should be 0
```
There are a few things about this cuke that can be improved, IMO:

* It's super low level and detailed, and doesn't function very well as documentation.  (Although, it is similar to some of our other cukes, but we have many that need to be cleaned up).
* Having test coverage only in a cuke means `rspec spec` (which is the main thing we use locally) can't be used to tell us if this feature is still working.

Putting these together, here's what I'd like to see:

* This feature should be covered by specs, and not just the cukes.  You can add an integration spec to `spec/integration` if you want to test it in an integration style.
* The cuke here should really be focused on documentation, and not on exhaustively showing every edge case.  Here's what I'm thinking it needs:
  * A narrative fleshing out the reason behind this feature and how to use it.  It should include a note about how if you use this you probably want to add a `--require` option to `.rspec` to ensure the option is always set even if no spec files are loaded by the command.
  * A background setting up a common context for the entire file:
    * A `spec_helper` file that sets the flag.
    * A `.rspec` file that has `--require spec_helper` to demonstrate how we recommend users set things up
    * A spec file not matching the normal spec file pattern (e.g. `some.spec` instead of `some_spec.rb`).
  * In one scenario, run `rspec` (with no options) and assert `Then it should fail with "0 examples, 0 failures"`.
  * I a 2nd scenario, run `rspec --pattern spec/**/*.spec` so the spec file is matched, and assert `Then it should pass with 1 example, 0 failures`

That way, there are 2 short scenarios that are very easy to read and understand, a clear explanation of the reason for the feature and how to set it up, and not much else to distract from that core documentation purpose.

(Also, this new file needs to be added to `features/.nav`).

Thoughts?

xmik replied:
Initially I wanted to add a test similar to https://github.com/rspec/rspec-core/blob/master/spec/rspec/core/runner_spec.rb#L431 , but I couldn't make it work: I got `LoadError` when adding:
```ruby
context "when fail_if_no_examples option is set" do
  include_context "spec files"

  it "returns 0 if examples pass and fail_if_no_examples is true" do
    runner = build_runner failing_spec_filename, "fail_if_no_examples", "true"
    expect(runner.run(err, out)).to eq 0
  end
end
```
It seemed appropriate place for such test, but since I cannot make it work and you suggest writing `spec/integration`, I'll provide such tests.

myronmarston commented on lib/rspec/core/runner.rb:
```diff
@@ -110,6 +110,9 @@ def setup(err, out)
         @configuration.reporter.report(@world.example_count(example_groups)) do |reporter|
           @configuration.with_suite_hooks do
+            if @world.example_count(example_groups) == 0 && @configuration.fail_if_no_examples
+              return 1
```
I suspect that the build is going to fail due to this line not being covered by any specs, but if you follow my advice above about adding specs for this it should solve that issue.

xmik commented:
I updated the PR with your remarks.

JonRowe commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,46 @@
+Feature: fail if no examples
+
+  Use the `fail_if_no_examples` option to make RSpec exit with exit status 1
```
Technically RSpec will exit with the configured error failure status, so prehaps this should read:
```
to make RSpec exit with a failure status (by default 1)
```
which also avoids the double exit :)

JonRowe commented on lib/rspec/core/runner.rb:
```diff
@@ -108,8 +108,12 @@ def setup(err, out)
+        @configuration.reporter.report(examples_count) do |reporter|
           @configuration.with_suite_hooks do
+            if examples_count == 0 && @configuration.fail_if_no_examples
+              return 1
```
This should be `@configuration.failure_exit_code`, we don't want to hard code 1.

xmik replied:
Corrected that and added unit test when `failure_exit_code` and `fail_if_no_examples` are both set :)

myronmarston commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,46 @@
+  This option may be particularly useful when you happen to not run RSpec tests
+  locally, but rely on CI to do this. This prevents from false positive builds,
+  when you expected some RSpec examples to be run, but none were run. Such
+   a situation may be caused by your misconfiguration or regression/major changes
```
The `a` is indented an extra space here...

myronmarston commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,46 @@
+        end
+      end
+      """
+    When I run `rspec --pattern spec/**/*_spec.rb`
```
The pattern you've named here isn't necessary.  It's the default pattern.  My suggestion about passing pattern was to pass a pattern that matches the file you defined in the background, e.g. `--pattern spec/**/*.spec`.  If you do that, there's no need to define a spec file in this scenario -- it can just be a short `When` and a `Then`.  OTOH, if you want to keep the step here defining a spec file, you can remove `--pattern` and remove the step in the background that defines the spec file as it doesn't really serve a purpose.

myronmarston commented on spec/support/aruba_support.rb:
```diff
@@ -10,7 +10,7 @@ module ArubaLoader
   let(:stdout) { StringIO.new }
 
-  attr_reader :last_cmd_stdout, :last_cmd_stderr
+  attr_reader :last_cmd_stdout, :last_cmd_stderr, :last_cmd_exitstatus
```
I'd prefer `last_cmd_exit_status` to `last_cmd_exitstatus`.  "Exit" and "status" are separate words.

xmik replied:
I was influenced by [`$?.exitstatus`](http://ruby-doc.org/core-2.2.0/Process/Status.html), but `last_cmd_exit_status` sounds better to me too.

myronmarston commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,46 @@
+Feature: fail if no examples
+
+  Use the `fail_if_no_examples` option to make RSpec exit with exit status 1
+  if there are 0 examples. Using this option, it is recommended to add:
```
I don't think the colon (`:`) on the end of this line is needed.

myronmarston commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,46 @@
+  locally, but rely on CI to do this. This prevents from false positive builds,
+  when you expected some RSpec examples to be run, but none were run. Such
+   a situation may be caused by your misconfiguration or regression/major changes
+  in RSpec.
```
You hand-wrapped this narrative nicely, which would usually be what we want...but there's a rendering issue on relish.  It inserts line breaks wherever we have them in these paragraphs, which means that when it renders the lines don't look even since it's not using a fixed-width font like it renders here on github.  To make it look nice, we usually remove the hand-wrapping, making each paragraph one long line.

Do you mind doing that?

myronmarston commented:
@xmik this is looking really close and is very well done!

xmik commented:
Thanks @myronmarston and @JonRowe for all the remarks. I incorporated them. It's nice to see that you care for RSpec and keep the code clean. Let me know if you want me to squash commits or correct something else.

myronmarston commented on features/configuration/fail_if_no_examples.feature:
```diff
@@ -0,0 +1,31 @@
+Feature: fail if no examples
+
+  Use the `fail_if_no_examples` option to make RSpec exit with a failure status (by default 1) if there are 0 examples. Using this option, it is recommended to add a `--require spec_helper` option to `.rspec` file to ensure the `fail_if_no_examples` option is set even if no spec files are loaded.
```
Suggestion: s/if there are 0 examples/if there are no examples/.

I think it reads more naturally and aligns with the config option name.

myronmarston commented:
> Let me know if you want me to squash commits or correct something else.

Please do squash it, then I'll gladly merge it :).

xmik commented:
` s/if there are 0 examples/if there are no examples/` incorporated and commits squashed.

JonRowe commented:
Thanks @xmik
